designs/isolate-cluster-upgrades.md

```diff
@@ -0,0 +1,221 @@
+* If a child object reference is changed to a different object, `Cluster` object's `generation` will change, increasing the aggregated cluster config's generation.
+* If the reference is kept pointing to the same object, but the child object is changed, this will increase its `generation`, which also increases the aggregated cluster config's generation.
+$$aggregatedGeneration = \sum_{n=1}^{clusterConfigObjs} generation_n$$
+
+
+#### Hidden `generation` changes
+Although it covers most of the scenarios, this solution still leaves in edge case unsolved. While `generation` can only be increased for a certain object, by changing a child object reference to a different object we can make the aggregated generation go down, or even worse, stay the same. The latter is bad since we will identify that reconcile request as a no-op, hence no reacting to the changes. This can happen if a referenced gets changed to child object with a $generation = prevObjGeneration -1$. That decreased will be compensated by the increase in the `Cluster`'s generation.
+
+The most straight forward solution is to keep track of two generations: the `Cluster` 's generation and the aggregated generation for the child objects. 
+* If any reference changes, the  `Cluster` 's generation will increase.
+* If any child object changes, the aggregated child object generation will increase. Note that multiple child objects changing at once won't be a problem since their generation can only increase.
+
+So now the logic becomes:
+
+```mermaid
+stateDiagram-v2
+    [*] --> calculate
+
+	calculate: aggregatedGeneration = sum(child.generation)
+    reconcile: Reconcile cluster
+    observedGeneration: generation == observedGeneration and aggregatedGeneration == childrenObservedGeneration
+    updateGeneration: observedGeneration = generation
+    updateChild: childrenObservedGeneration = aggregatedGeneration
+    done: Reconciliation done?
+    calculate --> observedGeneration
+    state observedGeneration <<choice>>
+        observedGeneration --> reconcile: no
+        observedGeneration --> [*]: yes
+
+    reconcile --> done
+    state done <<choice>>
+        done --> updateGeneration: yes
+
+    updateGeneration  --> updateChild
+    updateChild  --> [*]
+```
+
+### API changes
+We propose adding two fields to the `Cluster` status:
+
+```go
+type ClusterStatus struct {
+	...
+	// observedGeneration represents the .metadata.generation the last time the
+	// cluster was successfully reconciled.
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// childrenObservedGeneration represents the sum of all cluster's linked
+	// objects's .metadata.generation the last time the cluster was successfully reconciled.
+	ChildrenObservedGeneration int64 `json:"childrenObservedGeneration,omitempty"`
```

Comment:
Is this field what tracks aggregate? What about always incrementing `generation` whenever a child object changes, would that remove the need to track this here? Not sure if I missed this above.

Reply:
The problem is `generation` is managed by kubernetes, we don't control its value.